libraries/abstractions/common_io/include/iot_sdio.h

```diff
@@ -0,0 +1,720 @@
+ *          reference to the SDIO card slot that the user is interested in. Thereafter, user
+ *          uses this handle to call other APIs to perform tasks associated with the target
+ *          card slot. For example, iot_sdio_check_card_presence_status(pxSdioHandle, true)
+ *          checks sdio card presence, and iot_sdio_io_write_extended(pxSdioHandle)
+ *          writes data to sdio card in the slot, etc.
+ *
+ * @limitation
+ *          SDIO HAL APIs are intended to be used in embedded systems where each sdio host
+ *          controller is connected to either a single hot swappable sido card slot or multiple
+ *          eSDIO devices in a Shared Bus Configuration (Refer to Section 7.3 of "SD Specification
+ *          Part E1 SDIO").
+ *
+ *          These APIs are not suitable for systems where a single sdio host controller supports
+ *          multiple hot swappable slots. With multiple hot swappable slots available on a single
+ *          host controller, it is not possible pre-determine which card slot sdio device will be
+ *          inserted, and know which card slot handle to request. Since this type of host controller
+ *          is very rare even in non-embedded world, it is not worth to add complexity to SDIO HAL
+ *          APIs to support it.
+ */
+
+#ifndef _IOT_SDIO_H_
+#define _IOT_SDIO_H_
+
+#include <stdint.h>
+#include <stdbool.h>
+
+
+/**
+ * @defgroup iot_sdio SDIO HAL APIs
+ * @{
+ */
+
+/**
+ * @brief The return codes for the functions in SDIO
+ */
+ /*!< SDIO operation completed successfully. */
+#define IOT_SDIO_SUCCESS                             ( 0 )
+/*!< SDIO operation failed. */
+#define IOT_SDIO_FAIL                                ( 1 )
+/*!< SDIO host controller initialization failed. */
+#define IOT_SDIO_HOST_INIT_FAIL                      ( 2 )
+/*!< At least one parameter is invalid. */
+#define IOT_SDIO_INVALID_VALUE                       ( 3 )
+/*!< Sending command to or receive response from
+     card failed. */
+#define IOT_SDIO_TRANSFER_FAIL                       ( 4 )
+/*!< If the card is initialized in 3.3V signaling,
+     and card support high speed mode (SHS ==1),
+     host can issue CMD52 in RAW mode, setting EHS
+     bit in CCCR to one to switch to high speed mode
+     . This error code indicates that the sdio card
+     supports high speed mode but failed to switch to
+     high speed mode. The bus clock rate remains
+     unchanged in this case. */
+#define IOT_SDIO_SWITCH_HIGH_SPEED_FAIL              ( 5 )
+/*!< If the card is initialized in 3.3V signaling,
+     and card support high speed mode (SHS ==1),
+     host can issue CMD52 in RAW mode, setting EHS
+     bit in CCCR to one to switch to high speed mode
+     . This error code indicates that the sdio card
+     does not supports high speed mode. The bus clock
+     rate will be set to the default 25MHz in this
+     case. */
+#define IOT_SDIO_HIGH_SPEED_MODE_NOT_SUPPORTED       ( 6 )
```

Comment:
We can replace this with a generic "NOT SUPPORTED" error code.

Reply:
This is needed since it is one of the specific reasons why switch speed could fail.  This is different from API not supported on certain platform implementation.  I will add IOT_SDIO_NOT_SUPPORTED for that case.

Reply:
No, in iot_sdio_ioctl, when some request is not supported, we return generic "NOT SUPPORT". This eSDIOSwitchToHighSpeed also applies.

I can understand IOT_SDIO_SWITCH_HIGH_SPEED_FAIL is a specific one for high speed mode but "NOT SUPPORT" is not

Reply:
How do I distinguish the following:
1.  Older SD controller don't support high speed at all.
2. Newer SD controller support high speed, but the SDIO Card inserts doesn't support high speed.
3. For whatever reason, the implementer decide to not to implement this API and simply want to return "NOT_SUPPORTED"

Do I use the same return code for all 3 scenarios?  It don't seem to help debugging very much.

Reply:
I see your point now. I didn't think from the point of debugging. 
My opinion is that all the 3 cases are same to users. Do users need to distinguish or debug? 
I will let you make the decision.

Reply:
I will change "IOT_SDIO_HIGH_SPEED_MODE_NOT_SUPPORTED" to "IOT_SDIO_LOW_SPEED_CARD"